h2o-algos/src/main/java/hex/glm/GLM.java

```diff
@@ -2126,6 +2134,160 @@ private void fitIRLSM(Solver s) {
+          } else {
+            if (_betaConstraintsOn) // apply beta constraints without LS
+              bc.applyAllBounds(betaCnd);
+            Log.info(LogMsg("computed in " + (t2 - t1) + "+" + (t3 - t2) + "=" + (t3 - t1) + "ms, step = " + 1 + ((_lslvr != null) ? ", l1solver " + _lslvr : "")));
+          }
+
+          // Dispersion estimation part
+          if (negativebinomial.equals(_parms._family)){
+            converged = updateNegativeBinomialDispersion(iterCnt, _state.beta(), previousLLH, weights, response) && converged;
+          }
+          if (Math.abs(previousLLH - gram.likelihood) < _parms._objective_epsilon)
+            sameLLH ++;
+          else
+            sameLLH = 0;
+          converged = converged || sameLLH > 10;
+          previousLLH = gram.likelihood;
+        }
+      } catch (NonSPDMatrixException e) {
+        Log.warn(LogMsg("Got Non SPD matrix, stopped."));
+        throw e; // TODO: Decide if we want to return a bad model or no model at all.
+      }
+    }
+
+    private void updateTheta(double theta){
+      if (_state._glmw != null) {
+         _state._glmw._theta = theta;
+         _state._glmw._invTheta = 1./theta;
+      }
+      _parms._theta = theta;
+      _parms._invTheta =1./theta;
+      _model._parms._theta = theta;
+      _model._parms._invTheta = 1./theta;
+    }
+
+    private boolean updateNegativeBinomialDispersion(int iterCnt, double[] betaCnd, double previousNLLH, Vec weights, Vec response) {
+      double delta;
+      double theta;
+      boolean converged = false;
+      if (iterCnt == 1) {
+        theta = estimateNegBinomialDispersionMomentMethod(_model, betaCnd, _dinfo, weights, response);
+      } else {
+        theta = _parms._theta;
+        DispersionTask.GenPrediction gPred = new DispersionTask.GenPrediction(betaCnd, _model, _dinfo).doAll(
+                1, Vec.T_NUM, _dinfo._adaptedFrame);
+        Vec mu = gPred.outputFrame(Key.make(), new String[]{"prediction"}, null).vec(0);
+
+        NegativeBinomialGradientAndHessian nbGrad = new NegativeBinomialGradientAndHessian(theta).doAll(mu, response, weights);
+        delta = _parms._dispersion_learning_rate * nbGrad._grad / nbGrad._hess;
+        double bestLLH = Math.max(-previousNLLH, nbGrad._llh);
+        double bestTheta = theta;
+
+        delta = Double.isFinite(delta) ? delta : 1; // NaN can occur in extreme datasets so try to get out of this neighborhood just by linesearch
+
+        // Golden section search for the optimal size of delta
+        // Set lowerbound to -10 or lowest value that will keep theta > 0 which ever is bigger
+        // Negative value here helps with datasets where we use to diverge, I'm not sure yet if it's caused by some
+        // numerical issues or if the likelihood can get multimodal for some cases.
+        double lowerBound = (theta + 10 * delta < 0) ? (1 - 1e-15) * theta / delta : -10;
```

Comment:
you know that theta will never be negative.  There is no point trying to make it to be negative number.  I will set it to zero if it does not create NaN or Inf problems.  Otherwise, to something small like 1e-12 or something.

Reply:
The lower and upper bound are related to a multiplicative factor of `delta`, I'll call it `a` here:
`new_theta = old_theta - a*delta`.

So making `a` smaller than zero is still possible while keeping `theta > 0`. But it doesn't make much sense why going in opposite direction than the gradient could help but it does. I didn't investigate it much but sometimes there was really picked a negative `a` (i.e. going against the gradient had better log-likelihood). The cases where that happened often (maybe always) had extreme values in the `hex.glm.GLMModel.GLMWeightsFun#computeWeights` so my guess is that it can somehow correct smaller divergencies caused by some numerical issues.

And since I test on not very nice datasets as well the change from `lowerBound = -10` to `lowerBound = 0` is very visible on the results (left is with -10, right side is with 0).
![Screen Shot 2023-01-30 at 10 32 57](https://user-images.githubusercontent.com/61695433/215439936-26829faa-00ea-4b2c-a496-d5bb9302e63e.png)